### PR TITLE
Classification of the ANDI dataset using the Global Filter Network Transformer - s4717300

### PR DESCRIPTION
**Pull request by BenjaminJorgensen**
Author: Benjamin Jorgensen - s4717300

This project implements a Global Filter Vision Transformer (GFNet) for detecting Alzheimer's disease from MRI scans using the ANDI dataset. The model achieved a classification accuracy of 78.01% on the test set after 230 epochs. The result is only 1.99% less than the target of 80%.

The key component of the GFNet is transformation of the sample space into frequency space using the "Fast Fourier Transform" algorithm. The frequency vector is then used as the key matrix in the self-attention layer which is faster than typical n^2 self attention. This provides faster training and better extraction of features in small datasets such as ANDI.

---
The directory structure is as follows: 
`train.py` - Trains the model based on hyper-parameters given in `utils.py`

`predict.py` - Runs inference on a single file or evaluation on a test set.

`dataset.py` - Loads the dataset for training or inference/prediction

`modules.py` - Defines the GFNet architecture 

`utils.py` - Contains an Environment class that supplies hyper-parameters and other miscellaneous training variables

---
The model can perform inference on an image, classifying it into one of the two categories via the command
```
python predict.py ./models/GFNet.pth ./image.jpg
```

Training the model can be performed via the command
```
python train.py ./path_to_dataset
```
Further explanation of the usage can be found in the README.

Thank you for considering my pull request.

### CODE REVIEW

**Comment by wangzhaomxy** [comment]
**This is an initial inspection, no action is required at this point.**

**File Organizing:** Well-organized files.

**Problem Solving:**

- The algorithm solves the problem appropriately. 
- Accuracy in testing dataset: 0.74. 

**Model and functions:**

- It correctly uses PyTorch to construct the GFNet models and functions.

**Code design:** Good.

**Code comment and docstring:**

- Good code comments 
- Good function docstrings

**Difficulty:** Easy.

**Other Comments:**

- Good global filter visualisation.
- Sufficient commits
- Good ReadMe design

**Comment by BenjaminJorgensen** [comment]
Hi Zhao, 
Thank you for your review on my pull request.

I can't help but notice that you've incorrectly reported my testing accuracy to be 74% when the model is capable of a 78% accuracy as outlined on the fourth line of the README, in my pull request comment, and as can be seen in the README model accuracy graph.

I believe that this figure may have been mistakenly taken from the example output of performing inference with `predict.py` in the README. In case this wasn't clear, the example output was to demonstrate the format of the model output and isn't meant to be reflective of the model's maximum performance. Please keep this in mind when you apply the final marking scheme. 

Thank you and best of luck with your marking.

**Comment by wangzhaomxy** [comment]
> Hi Zhao, Thank you for your review on my pull request.
> 
> I can't help but notice that you've incorrectly reported my testing accuracy to be 74% when the model is capable of a 78% accuracy as outlined on the fourth line of the README, in my pull request comment, and as can be seen in the README model accuracy graph.
> 
> I believe that this figure may have been mistakenly taken from the example output of performing inference with `predict.py` in the README. In case this wasn't clear, the example output was to demonstrate the format of the model output and isn't meant to be reflective of the model's maximum performance. Please keep this in mind when you apply the final marking scheme.
> 
> Thank you and best of luck with your marking.

Hi Benjamin,

Thanks for your feedback.

I did realise the 78% acc in your report. However, the Acc figure in the Result part of your report shows "Model acc during training", which suspected indicates this acc is the training acc. I would give you 78% acc if you can clarify this part.

Cheers,
Zhao

**Comment by BenjaminJorgensen** [comment]
Ah, sorry for being unclear.
I described it as "during training" because the data was collected before the model had finished training. You are correct in that it has nothing to do with training at all and I can see how it would be misleading.
Thank you for pointing it out, I'll amend it shortly.
Thank you again!

**Comment by aniketgupta17** [comment]
Observational Feedback 

Pull Request:
Correctly created the Pull request from Topic Recognition Branch .
The pull request includes a clear description, summarizing the changes made in this submission.

File Organizing: Well-organized files.


Commit Log:
Commit messages are meaningful and provide insights into the code updates and progress made.
Commits demonstrate a logical development flow, covering different stages like data loading, model training, and testing.
Commits are regularly made, showing incremental development rather than bulk updates.

Documentation:
The README file provides a clear explanation of the algorithm and the problem it solves.
Code comments and docstrings are comprehensive, aiding readability and usability.
Proper GitHub markdown formatting is used, with organized headings, lists, and code blocks.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. Uses Data Aug</td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  GFNet (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). </td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>0</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Good work!
